## LCM
LCM looks for closed itemset with respect to an input minimum support

#### load the chess dataset

In [1]:
from skmine.datasets.fimi import fetch_chess
chess = fetch_chess()
chess.head()

0    [1, 3, 5, 7, 9, 11, 13, 15, 17, 19, 21, 23, 25...
1    [1, 3, 5, 7, 9, 12, 13, 15, 17, 19, 21, 23, 25...
2    [1, 3, 5, 7, 9, 12, 13, 16, 17, 19, 21, 23, 25...
3    [1, 3, 5, 7, 9, 11, 13, 15, 17, 20, 21, 23, 25...
4    [1, 3, 5, 7, 9, 11, 13, 15, 17, 19, 21, 23, 25...
Name: chess, dtype: object

In [2]:
chess.shape

(3196,)

#### fit_discover()
fit_discover makes pattern discovery more user friendly by outputting pretty formatted
patterns, instead of the traditional tabular format used in the `scikit` community

In [3]:
from skmine.itemsets import LCM
lcm = LCM(min_supp=2000, n_jobs=4)
# minimum support of 2000, running on 4 processes
%time patterns = lcm.fit_discover(chess)

CPU times: user 527 ms, sys: 353 ms, total: 880 ms
Wall time: 9.45 s


In [4]:
patterns.shape

(68967, 2)

This format in which patterns are rendered makes post hoc analysis easier

Here we filter patterns with a length strictly superior to 3

In [6]:
patterns[patterns.itemset.map(len) > 3]

itemset support
14         (25, 58, 17, 29)    2179
17          (3, 58, 17, 29)    2272
18          (25, 58, 3, 29)    2520
22         (25, 29, 58, 31)    2220
23          (29, 58, 3, 31)    2360
...                     ...     ...
68958       (52, 54, 5, 40)    2034
68959   (29, 54, 5, 40, 52)    2027
68964      (70, 52, 40, 29)    2001
68965      (70, 58, 40, 29)    2006
68966  (29, 70, 58, 40, 52)    2000

[67716 rows x 2 columns]

`Note`

Even when setting a very high minimum support threshold, we discovered more than 60K from only 3196 original transactions.
This is a good illustration of the so-called **pattern explosion problem**

------------
We could also get the top-k patterns in terms of supports, with a single line of code

In [8]:
patterns.nlargest(10, columns=['support'])  # top 10 patterns

itemset  support
0             (58)     3195
3749          (52)     3185
1118      (52, 58)     3184
4498          (29)     3181
8         (29, 58)     3180
3757      (52, 29)     3170
4506          (40)     3170
88        (40, 58)     3169
1126  (52, 29, 58)     3169
3800      (40, 52)     3159

---------------
#### fit_transform()
LCM can also be used as a preprocessing step in a pipeline working on tabular data

In [12]:
lcm = LCM(min_supp=2000, n_jobs=4)
lcm.fit_transform(chess)

(58)  (52)  (52, 58)  (29)  (29, 58)  (52, 29)  (40)  (40, 58)  \
0        1     1         1     1         1         1     1         1   
1        1     1         1     1         1         1     1         1   
2        1     1         1     1         1         1     1         1   
3        1     1         1     1         1         1     1         1   
4        1     1         1     1         1         1     1         1   
...    ...   ...       ...   ...       ...       ...   ...       ...   
3191     1     1         1     1         1         1     1         1   
3192     1     1         1     1         1         1     1         1   
3193     1     1         1     1         1         1     1         1   
3194     1     1         1     0         0         0     1         1   
3195     1     1         1     0         0         0     1         1   

      (52, 29, 58)  (52, 40)  ...  (17, 34, 9, 52, 60, 48, 58)  \
0                1         1  ...                            1   
1                1         1  ...                            1   
2                1         1  ...                            1   
3                1         1  ...                            1   
4                1         1  ...                            1   
...            ...       ...  ...                          ...   
3191             1         1  ...                            0   
3192             1         1  ...                            0   
3193             1         1  ...                            0   
3194             0         1  ...                            0   
3195             0         1  ...                            0   

      (40, 7, 46, 42, 58, 62)  (34, 9, 44, 52, 56, 29, 36, 58, 62)  \
0                           1                                    1   
1                           1                                    1   
2                           1                                    1   
3                           0                                    1   
4                           1                                    1   
...                       ...                                  ...   
3191                        1                                    1   
3192                        1                                    1   
3193                        1                                    1   
3194                        0                                    0   
3195                        0                                    0   

      (5, 7, 31, 48, 36, 58, 62)  (40, 34, 5, 52, 29, 31, 60, 42, 66, 58)  \
0                              1                                        1   
1                              1                                        1   
2                              1                                        1   
3                              1                                        1   
4                              1                                        1   
...                          ...                                      ...   
3191                           0                                        0   
3192                           0                                        0   
3193                           0                                        0   
3194                           0                                        0   
3195                           0                                        0   

      (40, 34, 60, 64, 29, 48, 36, 3, 58, 62)  (40, 52, 60, 54, 48, 58)  \
0                                           1                         1   
1                                           1                         1   
2                                           1                         1   
3                                           1                         1   
4                                           1                         1   
...                                       ...                       ...   
3191                                        0            